

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Articles &mdash; Deep Learning for Variant Calling 0.0.1 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Manuals" href="manuals.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Deep Learning for Variant Calling
          

          
          </a>

          
            
            
              <div class="version">
                0.0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="snakemake_rules.html">Snakemake Rules</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Scripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="methods.html">Methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="manuals.html">Manuals</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Articles</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Deep Learning for Variant Calling</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Articles</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/articles.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="articles">
<h1>Articles<a class="headerlink" href="#articles" title="Permalink to this headline">¶</a></h1>
<p>The articles consulted by this project and referenced in the report are the following:</p>
<ol class="arabic simple">
<li><ol class="upperalpha simple" start="6">
<li><p>Collins, E. Green, A. Guttmacher, et al. A vision for the future of genomics research. Nature, 422:835–847, 2003. URL <a class="reference external" href="https://doi.org/10.1038/nature01626">https://doi.org/10.1038/nature01626</a></p></li>
</ol>
</li>
<li><p>Carlson B. Snps-a shortcut to personalized medicine. Genet. Eng. Biotechnol. News, 28:12</p></li>
<li><ol class="upperalpha simple" start="19">
<li><p>Goodwin, J. McPherson, and W. McCombie. Coming of age: ten years of next-generation sequencing technologies. Nat Rev Genet, 17:333–351, 2016. URL <a class="reference external" href="https://doi.org/10.1038/nrg.2016.49">https://doi.org/10.1038/nrg.2016.49</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="18">
<li><p>Poplin, P. Chang, D. Alexander, et al. A universal snp and small-indel variant caller using deep neural networks. Nat Biotechnol, 36:983–987, 2018. URL <a class="reference external" href="https://doi.org/10.1038/nbt.4235">https://doi.org/10.1038/nbt.4235</a></p></li>
</ol>
</li>
<li><p>Christian Szegedy, Vincent Vanhoucke, Sergey Ioffe, Jonathon Shlens, and Zbiniew Wojna. Rethinking the inception architecture for computer vision. CoRR, abs/1512.00567, 2015. URL <a class="reference external" href="http://arxiv.org/abs/1512.00567">http://arxiv.org/abs/1512.00567</a></p></li>
<li><ol class="upperalpha simple" start="18">
<li><p>Luo, F.J. Sedlazeck, T. Lam, et al. A multi-task convolutional deep neural network for variant calling in single molecule sequencing. Nat Column, 10:998, 2019. URL <a class="reference external" href="https://doi.org/10.1038/s41467-019-09025-z">https://doi.org/10.1038/s41467-019-09025-z</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="18">
<li><p>Durbin, D. Altshuler, R. Durbin, et al. A map of human genome variation from population-scale sequencing. Nature, 467:1061–1073, 2010. URL <a class="reference external" href="https://doi.org/10.1038/nature09534">https://doi.org/10.1038/nature09534</a></p></li>
</ol>
</li>
<li><p>Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun. Delving deep into rectifiers: Surpassing human-level performance on imagenet classification. CoRR, abs/1502.01852, 2015. URL <a class="reference external" href="http://arxiv.org/abs/1502.01852">http://arxiv.org/abs/1502.01852</a></p></li>
<li><p>Günter Klambauer, Thomas Unterthiner, Andreas Mayr, and Sepp Hochreiter. Self-normalizing neural networks. In I. Guyon, U. V. Luxburg, S. Bengio, H. Wallach, R. Fergus, S. Vishwanathan, and R. Garnett, editors, Advances in Neural Information Processing Systems 30, pages 971–980. Curran Associates, Inc., 2017. URL <a class="reference external" href="http://papers.nips.cc/paper/6698-self-normalizing-neural-networks.pdf">http://papers.nips.cc/paper/6698-self-normalizing-neural-networks.pdf</a></p></li>
<li><ol class="upperalpha simple">
<li><p>McKenna, M. Hanna, E. Banks, et al. The genome analysis toolkit: a mapreduce framework for analyzing next-generation dna sequencing data. Genome Research, 20:1297–1303, 2010. URL <a class="reference external" href="https://doi.org/10.1101/gr.107524.110">https://doi.org/10.1101/gr.107524.110</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="5">
<li><p>Garrison and G. Marth. Haplotype-based variant detection from short-read sequencing. Arxiv, 2010. URL <a class="reference external" href="https://arxiv.org/abs/1207.3907">https://arxiv.org/abs/1207.3907</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="10">
<li><p>Zook, B. Chapman, J. Wang, et al. Integrating human sequence data sets provides a resource of benchmark snp and indel genotype calls. Nat Biotechnol, 32:246–251, 2014. URL <a class="reference external" href="https://doi.org/10.1038/nbt.2835">https://doi.org/10.1038/nbt.2835</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="10">
<li><p>Zook, D. Catoe, J. McDaniel, et al. Extensive sequencing of seven human genomes to characterize benchmark reference materials. Sci Data, 3, 2016. URL <a class="reference external" href="https://doi.org/10.1038/sdata.2016.25">https://doi.org/10.1038/sdata.2016.25</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="12">
<li><p>Heng, B. Handsaker, A. Wysoker, et al. The sequence alignment/map format and samtools. Bioinformatics, 25:2078–2079, 2009. URL <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="16">
<li><p>Danecek, A. Auton, Abecasis G., et al. The variant call format and vcftools. Bioinformatics, 27:2156–2158, 2011. URL <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btr330">https://doi.org/10.1093/bioinformatics/btr330</a></p></li>
</ol>
</li>
<li><ol class="upperalpha simple" start="10">
<li><p>Kster and S. Rahmann. Snakemakea scalable bioinformatics workflow engine. Bioinformatics, 28:2520–2522, 2012. URL <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bts480">https://doi.org/10.1093/bioinformatics/bts480</a></p></li>
</ol>
</li>
<li><p>Sepp Hochreiter and Jürgen Schmidhuber. Long short-term memory. Neural computation, 9(8):1735–1780, 1997</p></li>
<li><p>Kyunghyun Cho, Bart van Merrienboer, Çaglar Gülçehre, Fethi Bougares, Holger Schwenk, and Yoshua Bengio. Learning phrase representations using RNN encoder-decoder for statistical machine translation. CoRR, abs/1406.1078, 2014. URL <a class="reference external" href="http://arxiv.org/abs/1406.1078">http://arxiv.org/abs/1406.1078</a></p></li>
<li><ol class="upperalpha simple">
<li><p>Paszke, S. Gross, F. Massa, et al. Pytorch: An imperative style, high-performance deep learning library. In Advances in Neural Information Processing Systems 32, pages 8024–8035. Curran Associates, Inc., 2019. URL <a class="reference external" href="http://papers.neurips.cc/paper/9015-pytorch-an-imperative-style-high-performance-deep-learning-library.pdf">http://papers.neurips.cc/paper/9015-pytorch-an-imperative-style-high-performance-deep-learning-library.pdf</a></p></li>
</ol>
</li>
<li><p>Nitish Srivastava, Geoffrey Hinton, Alex Krizhevsky, et al. Dropout: A simple way to prevent neural networks from overfitting. Journal of Machine Learning Research, 15(56):1929–1958, 2014. URL <a class="reference external" href="http://jmlr.org/papers/v15/srivastava14a.html">http://jmlr.org/papers/v15/srivastava14a.html</a></p></li>
</ol>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="manuals.html" class="btn btn-neutral float-left" title="Manuals" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Daniel Biro

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>